src/App.tsx

```diff
@@ -1,24 +1,260 @@
+import { TodoHeader } from './components/TodoHeader/TodoHeader';
 
 export const App: React.FC = () => {
-  if (!USER_ID) {
-    return <UserWarning />;
-  }
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [filterCase, setFilterCase] = useState(FilterCase.all);
+  const [error, setError] = useState('');
+  const [isLoading, setIsLoading] = useState(false);
+  const [loadingTodoIds, setLoadingTodoIds] = useState<number[]>([]);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+
+  const newTodoField = useRef<HTMLInputElement>(null);
+
+  useEffect(() => {
+    if (newTodoField.current) {
+      newTodoField.current.focus();
+    }
+  }, [isLoading]);
+
+  useEffect(() => {
+    setError('');
+
+    getTodos()
+      .then(setTodos)
+      .catch(() => {
+        setError('Unable to load todos');
+      });
+  }, []);
+
+  const timerId = useRef<number>(0);
+
+  useEffect(() => {
+    if (timerId.current) {
+      window.clearTimeout(timerId.current);
+    }
+
+    timerId.current = window.setTimeout(() => {
+      setError('');
+    }, 3000);
+  }, [error]);
+
+  const handleAddTodo = (todoTitle: string) => {
+    setTempTodo({
+      id: 0,
+      title: todoTitle.trim(),
+      userId: 0,
+      completed: false,
+    });
+
+    setIsLoading(true);
+
+    return addTodo(todoTitle)
+      .then((newTodo) => {
+        setTodos((prevTodos) => [...prevTodos, newTodo]);
+
+        newTodoField.current?.focus();
+      })
+      .catch(() => {
+        setError('Unable to add a todo');
+        throw new Error();
+      })
+      .finally(() => {
+        setTempTodo(null);
+        setIsLoading(false);
+      });
+  };
+
+  const handleDeleteTodo = (todoId: number) => {
+    setLoadingTodoIds((prevLoadingTodoIds) => [...prevLoadingTodoIds, todoId]);
+    setIsLoading(true);
+
+    deleteTodo(todoId)
+      .then(() => {
+        setTodos((prevTodos) => prevTodos.filter((todo) => todo.id !== todoId));
+
+        newTodoField.current?.focus();
+      })
+      .catch(() => {
+        setError('Unable to delete a todo');
+      })
+      .finally(() => {
+        setLoadingTodoIds((prevTodoIds) => prevTodoIds
+          .filter((id) => id !== todoId));
+        setIsLoading(false);
+      });
+  };
+
+  const handleRenameTodo = (todo: Todo, newTodoTitle: string) => {
+    setLoadingTodoIds((prevLoadingTodoIds) => [...prevLoadingTodoIds, todo.id]);
+
+    updateTodo({
+      id: todo.id,
+      title: newTodoTitle,
+      userId: todo.userId,
+      completed: todo.completed,
+    })
+      .then((updatedTodo) => {
+        setTodos((prevTodos) => prevTodos.map((currentTodo) => (
+          currentTodo.id !== updatedTodo.id ? currentTodo : updatedTodo
+        )));
+      })
+      .catch(() => {
+        setError('Unable to update a todo');
+      })
+      .finally(() => {
+        setLoadingTodoIds((prevTodoIds) => prevTodoIds
+          .filter((id) => id !== todo.id));
+        setIsLoading(false);
+      });
```

Comment:
You use similar logic to update the status there, so it can be a single handler that takes as an argument the data to be updated.